### PR TITLE
refactor ack consumer using admin api

### LINKED ISSUE #229: [BUG] the topic can not remove after the producer and consumer finished
**Describe the bug**
the topic can not remove after the producer and consumer finished.

**To Reproduce**
1, create a partitioned topic
`pulsar-admin persistent create-partitioned-topic t-kop/ns-kop/lookup2 --partitions 30`
2, produce `100` message using the created topic
3, consume all produced messages.
4, delete the used topic `lookup2` with pulsar command
`./bin/pulsar-admin topics delete-partitioned-topic  t-kop/ns-kop/lookup2`
5, See error:
`Failed to perform http delete request: javax.ws.rs.ClientErrorException: HTTP 412 Precondition Failed
Topic has active producers/subscriptions`

**Expected behavior**
the topic can remove after the producer and consumer finished.

**Screenshots**
image

### PR DESCRIPTION
**Pull request by dockerzhang**
Fixes #229 

refactor ack consumer using admin API

### CODE REVIEW

**Comment by BewareMyPower** [comment]
Tips, add `Fixes #xxx ` in the first line, then the PR will be linked to the issue `xxx`.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java [hunk 1]
```diff
@@ -86,7 +86,7 @@ public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exceptio
             // set auto read to true if channel is writable.
             ctx.channel().config().setAutoRead(true);
         } else {
-            log.warn("channel is not writable, disable auto reading for back pressing");
+            log.debug("channel is not writable, disable auto reading for back pressing");
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java:89]: The change that is unrelated to the PR should be put in a new PR, no matter how minor it is.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java [hunk 2]
```diff
@@ -930,7 +930,7 @@ private void handleListOffsetRequestV0(KafkaHeaderAndRequest listOffset,
 
         // in v0, the iterator is offsetData,
         // in v1, the iterator is partitionTimestamps,
-        log.warn("received a v0 listOffset: {}", request.toString(true));
+        log.debug("received a v0 listOffset: {}", request.toString(true));
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:933]: The same as I said before.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java (hunk 2 above)
Comment (BewareMyPower) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:933]: By the way, if a new PR was pushed to change log level, `log.debug` should be wrapped in a `if (log.isDebugEnabled())` block. And

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java (hunk 1 above)
Comment (dockerzhang) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java:89]: Lots of useless warn log when testing, I will fix it in another issue.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java (hunk 2 above)
Comment (dockerzhang) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:933]: @BewareMyPower  there is a UT failure, do you know why it's caused by this PR?

Comment (dockerzhang) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java:933]: I think the failed UT is caused by no consumer `backlog ` metrics if using `resetCursorAsync`.
 we should think about how to solve the consumer metrics problem at the same time.

**Comment by dockerzhang** [comment]
@BewareMyPower I think the failed UT is caused by no consumer backlog metrics if using resetCursorAsync.
we should think about how to solve the consumer metrics problem at the same time.

**Comment by jiazhai** [comment]
@dockerzhang  @BewareMyPower  We added a consumer for the topic to track the real topic backlog. It is intended to use this consumer to keep the cursor position. 
In this change, if we change to use admin API, seems we add the reset cursor, but not provide the create/init cursor.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/OffsetAcker.java [hunk 3]
```diff
@@ -66,57 +52,8 @@ public void ackOffsets(String groupId, Map<TopicPartition, OffsetAndMetadata> of
...
-                }
-                try {
-                    consumer.close();
-                } catch (Exception e) {
-                    log.warn("Error when close consumer topic: {}, sub: {}.",
-                        consumer.getTopic(), consumer.getSubscription(), e);
-                }
-            });
+            MessageId messageId = MessageIdUtils.getMessageId(offsetAndMetadata.offset());
+            pulsarAdmin.topics().resetCursorAsync(topicPartition.toString(), groupId, messageId);
```
Comment (jiazhai) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/OffsetAcker.java:56]: seems not find the place to init this cursor

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/OffsetAcker.java [hunk 4]
```diff
@@ -14,48 +14,34 @@
...
-    // map off consumser: <groupId, consumers>
-    Map<String, Map<TopicPartition, CompletableFuture<Consumer<byte[]>>>> consumers = new ConcurrentHashMap<>();
-
     public void addOffsetsTracker(String groupId, byte[] assignment) {
         ByteBuffer assignBuffer = ByteBuffer.wrap(assignment);
         Assignment assign = ConsumerProtocol.deserializeAssignment(assignBuffer);
         if (log.isDebugEnabled()) {
             log.debug(" Add offsets after sync group: {}", assign.toString());
         }
-        assign.partitions().forEach(topicPartition -> getConsumer(groupId, topicPartition));
```
Comment (jiazhai) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/OffsetAcker.java:58]: maybe here we need to init/add the original cursor to keep the backlog position.

**Review by jiazhai — COMMENTED** [review]
> @BewareMyPower I think the failed UT is caused by no consumer backlog metrics if using resetCursorAsync.
> we should think about how to solve the consumer metrics problem at the same time.

@dockerzhang we need to keep this UT passed, it test the real backlog.

**Comment by jiazhai** [comment]
In the original code, we need a chance to remove the added consumer in offsetAcker, e.g. at the time the consumer group removed.

**Comment by dockerzhang** [comment]
@jiazhai thanks for your suggestion, this issue is related to how to reuse pulsar consumer metrics in kop, and I will reconsider the problem.

**Comment by dockerzhang** [comment]
this issue will close because all connected consumers will close after the group is deleted using the Kafka client command.

The detail code is
https://github.com/streamnative/kop/blob/master/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupCoordinator.java#L673
